src/base/utils/version.h

```diff
@@ -0,0 +1,184 @@
+
+#include <array>
+#include <stdexcept>
+
+#include <QDebug>
+#include <QString>
+#include <QStringList>
+
+namespace Utils
+{
+
+    template <class T, std::size_t N, std::size_t Mandatory = N>
+    class Version
+    {
+        static_assert(N >= Mandatory,
+                      "The number of mandatory components may not be smaller than the total number of components");
+
+    public:
+        typedef T ComponentType;
+        typedef Version<T, N, Mandatory> ThisType;
+
+        constexpr Version()
+            : m_components {}
+        {
+        }
+
+        constexpr Version(const ThisType &other) = default;
+
+        template <class ... Other>
+        constexpr Version(Other ... components)
+            : m_components {{components ...}}
+        {
+        }
+
+        Version(const QString &version)
+            : Version {version.split(QLatin1Char('.'))}
+        {
+        }
+
+        Version(const QByteArray &version)
+            : Version {version.split('.')}
+        {
+        }
+
+        ComponentType majorNumber() const
+        {
+            static_assert(N >= 1, "Array size is too small");
+            return (*this)[0];
+        }
+
+        ComponentType minorNumber() const
+        {
+            static_assert(N >= 2, "Array size is too small");
+            return (*this)[1];
+        }
+
+        ComponentType revisionNumber() const
+        {
+            static_assert(N >= 2, "Array size is too small");
+            return (*this)[2];
+        }
+
+        ComponentType patchNumber() const
+        {
+            static_assert(N >= 3, "Array size is too small");
+            return (*this)[3];
+        }
+
+        ComponentType operator[](std::size_t i) const
+        {
+            return m_components.at(i);
+        }
+
+        operator QString() const
+        {
+            // find the last one non-zero component
+            std::size_t lastSignificantIndex = N - 1;
+            while (lastSignificantIndex >= 0 && (*this)[lastSignificantIndex] == 0)
+                --lastSignificantIndex;
+
+            if (lastSignificantIndex + 1 < Mandatory)     // lastSignificantIndex >= 0
+                lastSignificantIndex = Mandatory - 1;     // and Mandatory >= 1
+
+            QString res = QString::number((*this)[0]);
+            for (std::size_t i = 1; i <= lastSignificantIndex; ++i)
+                res += QLatin1Char('.') + QString::number((*this)[i]);
+            return res;
+        }
+
+        bool operator==(const ThisType &other) const
+        {
+            return m_components == other.m_components;
+        }
+
+        bool operator<(const ThisType &other) const
+        {
+            return m_components < other.m_components;
+        }
+
+        bool operator>(const ThisType &other) const
+        {
+            return m_components > other.m_components;
+        }
+
+        template <class AQString>
+        static Version tryParse(const AQString &s, const Version &defaultVersion)
```

Comment:
Will leave it as it is then, because with the 'ok' parameters it would be needed to rewrite the parsing/constructor relation the other way round.